WcaOnRails/app/controllers/registrations_controller.rb

```diff
@@ -98,6 +108,101 @@ def export
+    new_locked_users = []
+    ActiveRecord::Base.transaction do
+      registrations.each do |registration|
+        user = user_for_registration!(registration, new_locked_users)
+        registration_competition_events_attributes = competition.competition_events.map do |competition_event|
+          registration[competition_event.event_id.to_sym] == "1" ? { competition_event_id: competition_event.id } : nil
+        end.compact
+        competition.registrations.create!(
+          user_id: user.id,
+          accepted_at: Time.now,
+          accepted_by: current_user.id,
+          registration_competition_events_attributes: registration_competition_events_attributes,
+        )
+      end
+    end
+    new_locked_users.each do |user|
+      RegistrationsMailer.notify_registrant_of_locked_account_creation(user, competition).deliver_later
+    end
+    redirect_to competition_registrations_import_url(competition)
+  rescue StandardError => error
+    flash[:danger] = error.to_s
+    redirect_to competition_registrations_import_url(competition)
+  end
+
+  private def user_for_registration!(registration, new_locked_users)
+    if registration[:wca_id].present?
+      user = User.find_by(wca_id: registration[:wca_id])
+      if user
+        if user.dummy_account?
+          email_user = User.find_by(email: registration[:email])
+          if email_user
+            if email_user.wca_id.present?
+              raise "There is already a user with email #{registration[:email]}"\
+                    ", but it has WCA ID of #{email_user.wca_id} instead of #{registration[:wca_id]}."
+            else
+              email_user.update!(wca_id: registration[:wca_id]) # User hooks will also remove the dummy user account.
+              email_user
+            end
+          else
+            user.skip_reconfirmation!
+            user.update!(dummy_account: false, email: registration[:email])
+            user
+          end
+        else
+          user # Use this account.
+        end
+      else
+        # Create a locked account with confirmed WCA ID.
+        create_locked_account!(registration).tap do |locked_user|
+          new_locked_users << locked_user
+        end
+      end
+    else
+      email_user = User.find_by(email: registration[:email])
+      # Use the user if exists, otherwise create a locked account without WCA ID.
+      email_user || create_locked_account!(registration).tap do |locked_user|
+        new_locked_users << locked_user
+      end
+    end
+  end
+
+  private def create_locked_account!(registration)
+    User.new(
+      name: registration[:name],
+      email: registration[:email],
+      wca_id: registration[:wca_id],
+      country_iso2: Country.c_find(registration[:country]).iso2,
+      gender: registration[:gender],
+      dob: registration[:birth_date],
+      encrypted_password: "",
+    ).tap do |user|
+      user.define_singleton_method(:password_required?) { false }
```

Comment:
[`define_singleton_method`](https://ruby-doc.org/core-2.5.1/Object.html#method-i-define_singleton_method) simply let's you add/override a method for the given object, with no impact on other objects from its class.` I found it [here](https://stackoverflow.com/a/45589123). The comment above that suggests to do it the other way, I just prefer not to add `attr_accessor`s for a one-time usage, if possible. I can comment this line to point to the SO answer, or use the other workaround, what do you prefer?

The `create_locked_account!` method accepts a very specific hash (a CSV row), but I'm ok with moving it to the `User` model. `User.new_locked_account` you suggested implies a new unsaved instance, I'd rather stick with `create_` to keep the saving stuff together (i.e. `define_singleton_method`, `skip_confirmation!`, `save!`).

Reply:
Ah, sorry I didn't explain my suggestion very well. I do agree that it's weird for the User model to know about the CSV row hash. I was picturing something more like:

```ruby
### WcaOnRails/app/controllers/registrations_controller.rb
def self.create_locked_account!(registration_csv_data)
  User.new_locked_account(
    name: registration_csv_data[:name],
    email: registration_csv_data[:email],
    wca_id: registration_csv_data[:wca_id],
    country_iso2: Country.c_find(registration_csv_data[:country]).iso2,
    gender: registration_csv_data[:gender],
    dob: registration_csv_data[:birth_date],
  ).save!
end

### models/user.rb
def self.new_locked_account(attributes = {})
  # A locked account is just a user with an empty password. It's impossible to sign into an account with an empty password, so the only way to log into a locked account is to reset its password.
  User.new(attributes.merge(encrypted_password: "")).tap do |user|
    user.define_singleton_method(:password_required?) { false }
    user.skip_confirmation!
  end
end
```

This may feel kind of boilerplate-y, but the thing I like about this is that it separates the "how do you create an account with no password?" from the "what data should go into this locked account?"

Reply:
I don't feel strongly about the `define_singleton_method` vs `attr_writer` approach. Maybe would be nice to link to the SO thread, but I also don't feel strongly about that. Whatever you think is best!